File: RxJavaTutorial/app/src/main/java/com/example/rxjavatutorial/MainActivity.java
package com.example.rxjavatutorial;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;

public class MainActivity extends AppCompatActivity {


    Button button;
    EditText editText;
    TextView textView;

    Observable<String> observable;
    Observer<String> observer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        editText = findViewById(R.id.editTextTextPersonName);
        button = findViewById(R.id.button);
        textView = findViewById(R.id.textView);

        createObservable();
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                observable.safeSubscribe(observer);

            }
        });
    }

    private void createObservable() {

        observable = Observable.create(new ObservableOnSubscribe<String>() {
            @Override
            public void subscribe(@NonNull ObservableEmitter<String> subscriber) throws Exception {

                subscriber.onNext(editText.getText().toString());
                subscriber.onComplete();

            }
        });

        observer = new Observer<String>() {
            @Override
            public void onSubscribe(@NonNull Disposable d) {

            }

            @Override
            public void onNext(@NonNull String s) {

                textView.setText(s);

            }

            @Override
            public void onError(@NonNull Throwable e) {

            }

            @Override
            public void onComplete() {

            }
        };
    }
}

File: ViewModel/app/src/main/java/com/example/viewmodel/MainActivity.kt
package com.example.viewmodel

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    lateinit var myViewModel : MyViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

          myViewModel = ViewModelProvider(this, MyViewModelFactory(10)).get(MyViewModel::class.java)
            tvIncrement.text = myViewModel.count.toString()


        incrementCount.setOnClickListener {

            myViewModel.increment()
            tvIncrement.text = myViewModel.count.toString()


        }
    }
}